wp-includes/sqlite-ast/class-wp-sqlite-driver.php

```diff
@@ -0,0 +1,2592 @@
+
+	/**
+	 * Class variable to store the result of the query.
+	 *
+	 * @access private
+	 *
+	 * @var array reference to the PHP object
+	 */
+	private $results = null;
+
+	/**
+	 * Class variable to check if there is an error.
+	 *
+	 * @var boolean
+	 */
+	public $is_error = false;
+
+	/**
+	 * Class variable to store the file name and function to cause error.
+	 *
+	 * @access private
+	 *
+	 * @var array
+	 */
+	private $errors;
+
+	/**
+	 * Class variable to store the error messages.
+	 *
+	 * @access private
+	 *
+	 * @var array
+	 */
+	private $error_messages = array();
+
+	/**
+	 * Class variable to store the affected row id.
+	 *
+	 * @var int integer
+	 * @access private
+	 */
+	private $last_insert_id;
+
+	/**
+	 * Number of rows found by the last SELECT query.
+	 *
+	 * @var int
+	 */
+	private $last_select_found_rows;
+
+	/**
+	 * Number of rows found by the last SQL_CALC_FOUND_ROW query.
+	 *
+	 * @var int integer
+	 */
+	private $last_sql_calc_found_rows = null;
+
+	/**
+	 * Class variable to store the number of rows affected.
+	 *
+	 * @var int integer
+	 */
+	private $affected_rows;
+
+	/**
+	 * Variable to emulate MySQL affected row.
+	 *
+	 * @var integer
+	 */
+	private $num_rows;
+
+	/**
+	 * Return value from query().
+	 *
+	 * Each query has its own return value.
+	 *
+	 * @var mixed
+	 */
+	private $return_value;
+
+	/**
+	 * Variable to keep track of nested transactions level.
+	 *
+	 * @var int
+	 */
+	private $transaction_level = 0;
+
+	/**
+	 * Value returned by the last exec().
+	 *
+	 * @var mixed
+	 */
+	private $last_exec_returned;
+
+	/**
+	 * The PDO fetch mode passed to query().
+	 *
+	 * @var mixed
+	 */
+	private $pdo_fetch_mode;
+
+	/**
+	 * @var WP_SQLite_Information_Schema_Builder
+	 */
+	private $information_schema_builder;
+
+	/**
+	 * Constructor.
+	 *
+	 * Create PDO object, set user defined functions and initialize other settings.
+	 * Don't use parent::__construct() because this class does not only returns
+	 * PDO instance but many others jobs.
+	 *
+	 * @param string $db_name The database name. In WordPress, the value of DB_NAME.
+	 * @param PDO|null $pdo The PDO object.
+	 */
+	public function __construct( string $db_name, ?PDO $pdo = null ) {
+		$this->db_name = $db_name;
+		if ( ! $pdo ) {
+			if ( ! is_file( FQDB ) ) {
+				$this->prepare_directory();
+			}
+
+			$locked      = false;
+			$status      = 0;
+			$err_message = '';
+			do {
+				try {
+					$options = array(
+						PDO::ATTR_ERRMODE           => PDO::ERRMODE_EXCEPTION,
+						PDO::ATTR_STRINGIFY_FETCHES => true,
+						PDO::ATTR_TIMEOUT           => 5,
+					);
+
+					$dsn = 'sqlite:' . FQDB;
+					$pdo = new PDO( $dsn, null, null, $options );
+				} catch ( PDOException $ex ) {
+					$status = $ex->getCode();
+					if ( self::SQLITE_BUSY === $status || self::SQLITE_LOCKED === $status ) {
+						$locked = true;
+					} else {
+						$err_message = $ex->getMessage();
+					}
+				}
+			} while ( $locked );
```

Comment:
Do we need a timeout to avoid an infinite loop here? Or, even better, could we trust PDO::ATTR_TIMEOUT and avoid the loop entirely?

Reply:
@adamziel I'll try to debug this somehow. It would be great to get rid of this loop altogether.